chain-storage/src/lib.rs

```diff
@@ -1,1079 +1,832 @@
-use crate::index::{ChainStorageIndex, IndexCreationError};
-use chain_core::property::{Block, BlockId, Serialize};
-use rusqlite::{types::Value, Connection, TransactionBehavior};
-use std::{
-    marker::PhantomData,
-    path::{Path, PathBuf},
-    sync::{Arc, RwLock},
-    time::Duration,
-};
-use thiserror::Error;
-
-mod index;
-
-#[derive(Debug, Error)]
-pub enum Error {
-    #[error("block not found")]
-    BlockNotFound,
-    // FIXME: add BlockId
-    #[error("database backend error")]
-    BackendError(#[from] Box<dyn std::error::Error + Send + Sync>),
-    #[error("Block already present in DB")]
-    BlockAlreadyPresent,
-    #[error("the parent block is missing for the required write")]
-    MissingParent,
+//! Block storage module.
+//!
+//! # Data model
+//!
+//! This storage is designed to be independent from a particular block strucure.
+//! At the same time, since it is designed specifically for block chains, it
+//! handles the minimum amount of data required to maintain consistency. Such
+//! data includes:
+//!
+//! * Block ID
+//! * ID of the parent block
+//! * Chain length
+//!
+//! This data is provided alongside a block in the `BlockInfo` structure.
+//!
+//! # Volatile + permanent storage model
+//!
+//! Since blockchain can be branching extensively, this library provides the
+//! mechanism to have multiple blockchain branches. However, doing so for all
+//! block screates a lot of overhead for maintaining volatile storage. At the
+//! same time, branches usually die out pretty fast.
+//!
+//! Given that, the storage is separated into two parts:
+//!
+//! * Volatile storage should be used for the part of a block chain, where a
+//!   developer may need access to different branches.
+//! * Permanent storage for the part of block chain, that is guaranteed not to
+//!   change anymore.
+//!
+//! ## Moving blocks to the permanent store.
+//!
+//! If you have this block structure and call
+//! `store.flush_to_permanent_store(block 2 id)`, then `Block 1` and `Block 2`
+//! will be moved to the permanent storage. If you call
+//! `store.flush_to_permanent_store(block 3 id)`, `Block 3` will also be moved
+//! to the permanent store, but `Block 3'` will still exist. Note that if you
+//! call `store.get_blocks_by_chain_length(3)` only `Block 3` (which is in the
+//! permanent store) will be returned; and you cannot call
+//! `store.flush_to_permanent_store(block 3' id)` now.
```

Comment:
Since Block 3 is not in the permanent storage, the fact that the _prime_ branch (`Block 3'` and `Block 4'`) are forking from a block that is already being dismissed should flush it out of the volatile storage: i.e. as soon a branch points to a block in the permanent storage that is below the last block of the permanent storage needs to be removed/pruned.

Reply:
Not sure if we want to do that automatically in the storage module. The reason for this is that we will need to scan all branches to check if they begin in the permanent storage. Instead of this we can just leave removing branches up to the application logic and this is most likely that the application will actually do that.